### PR TITLE
fix(ivy): ensure unique template and listener function identifiers

### PR DESCRIPTION
**Pull request by gkalpak**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

## PR Type
- [x] Bugfix

## What is the current behavior?
There can be identical template/listener function names for a component's template, if it has multiple similarly structured nested sub-templates or listeners.
This results in build errors: `Identifier '<SOME_IDENTIFIER>' has already been declared`

## What is the new behavior?
This PR fixes this by ensuring that the template index is included in the `contextName` passed to the `TemplateDefinitionBuilder` responsible for processing nested sub-templates. Similarly, the template or element index is included in the listener names.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Blocking: #26947

### CODE REVIEW

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/test/compliance/r3_view_compiler_template_spec.ts [hunk 1]
```diff
@@ -440,4 +440,120 @@ describe('compiler compliance: template', () => {
     expectEmit(result.source, template, 'Incorrect template');
 
   });
+
+  it('should create unigue template function names even for similar nested template structures', () => {
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/compliance/r3_view_compiler_template_spec.ts]: unigue -> unique

File: packages/compiler-cli/test/compliance/r3_view_compiler_template_spec.ts [hunk 2]
```diff
@@ -440,4 +440,120 @@ describe('compiler compliance: template', () => {
...
+
+    // Expected template function:
+    // - 5 for AComponent's template.
+    // - 9 for BComponent's template.
+    // - 2 for the two components.
+    expect(allTemplateFunctionsNames.length).toBe(5 + 9 + 2);
+    expect(allTemplateFunctionsNames).toEqual(uniqueTemplateFunctionNames);
+  });
+
+  it('should create unigue listener function names even for similar nested template structures', () => {
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/compliance/r3_view_compiler_template_spec.ts]: unique

**Review by petebacondarwin — APPROVED** [review]

File: packages/compiler-cli/test/compliance/r3_view_compiler_template_spec.ts (hunk 1 above)
Comment (gkalpak) [reply packages/compiler-cli/test/compliance/r3_view_compiler_template_spec.ts]: 😱

**Review by kara — APPROVED** [review]
LGTM

**Comment by kara** [comment]
[presubmit](http://test/OCL:226516414:BASE:226516456:1545420037852:67bef9ad)

**Comment by gkalpak** [comment]
Rebased on master.

**Pull request closed without merging**